aptos-move/framework/aptos-stdlib/sources/cryptography/bulletproofs.move

```diff
@@ -0,0 +1,402 @@
+/// This module implements a Bulletproof range proof verifier on the Ristretto255 curve.
+///
+/// A Bulletproof-based zero-knowledge range proof is a proof that a Pedersen commitment
+/// $c = v G + r H$ commits to an $n$-bit value $v$ (i.e., $v \in [0, 2^n)$). Currently, this module only supports
+/// $n \in \{8, 16, 32, 64\}$ for the number of bits.
+///
+/// Additionally, this module also supports range proofs over ElGamal ciphertexts, which can be viewed as a Pedersen
+/// commitment.
+module aptos_std::bulletproofs {
+    use std::features;
+    use aptos_std::elgamal;
+    use aptos_std::pedersen;
+    use aptos_std::ristretto255::{Self, RistrettoPoint};
+
+    //
+    // Constants
+    //
+
+    /// The maximum range supported by the Bulletproofs library is [0, 2^{64}).
+    const MAX_RANGE_BITS : u64 = 64;
+
+    //
+    // Error codes
+    //
+
+    /// There was an error deserializing the range proof.
+    const E_DESERIALIZE_RANGE_PROOF: u64 = 1;
+
+    /// The committed value given to the prover is too large.
+    const E_VALUE_OUTSIDE_RANGE: u64 = 2;
+
+    /// The range proof system only supports proving ranges of type $[0, 2^b)$ where $b \in \{8, 16, 32, 64\}$.
+    const E_RANGE_NOT_SUPPORTED: u64 = 3;
+
+    /// The native functions have not been rolled out yet.
+    const E_NATIVE_FUN_NOT_AVAILABLE: u64 = 4;
+
+    //
+    // Structs
+    //
+
+    /// Represents a zero-knowledge range proof that a value committed inside a Pedersen commitment lies in
+    /// `[0, 2^{MAX_RANGE_BITS})`.
+    struct RangeProof has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    //
+    // Public functions
+    //
+
+    /// Returns the maximum # of bits that the range proof system can verify proofs for.
+    public fun get_max_range_bits(): u64 {
+        MAX_RANGE_BITS
+    }
+
+    /// Deserializes a range proof from a sequence of bytes.
```

Comment:
serialization format, just to be complete?

Reply:
Fixed (will commit soon).